File: apps/guzo-api/src/main/java/et/guzo/web/controller/WalletController.java
package et.guzo.web.controller;

import et.guzo.common.ApiResponse;
import et.guzo.security.SecurityUtil;
import et.guzo.service.WalletService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.Map;

@RestController
@RequestMapping("/wallet")
@RequiredArgsConstructor
public class WalletController {

    private final WalletService walletService;

    @GetMapping
    public ApiResponse<Map<String, Object>> summary() {
        return ApiResponse.ok(walletService.summary(SecurityUtil.requireUser().getId()));
    }

    @GetMapping("/transactions")
    public ApiResponse<Map<String, Object>> transactions(
        @RequestParam(defaultValue = "1") int page,
        @RequestParam(defaultValue = "20") int limit
    ) {
        return ApiResponse.ok(walletService.transactions(SecurityUtil.requireUser().getId(), page, limit));
    }

    @PostMapping("/top-up")
    public ApiResponse<Map<String, Object>> topUp(@RequestBody Map<String, Object> body) {
        BigDecimal amount = body.get("amount") != null ? new BigDecimal(body.get("amount").toString()) : BigDecimal.ZERO;
        String description = body.get("description") != null ? body.get("description").toString() : null;
        return ApiResponse.ok(walletService.topUp(SecurityUtil.requireUser().getId(), amount, description), "Top-up successful");
    }
}

File: apps/server/src/modules/wallet/wallet.controller.ts
import type { Request, Response } from 'express';
import { walletService } from './wallet.service.js';
import { asyncHandler } from '../../utils/asyncHandler.js';
import { ok } from '../../utils/ApiResponse.js';
import { parseListQuery } from '../../utils/pagination.js';

export const walletController = {
  summary: asyncHandler(async (req: Request, res: Response) => {
    const data = await walletService.summary(req.user!.id);
    return ok(res, data, 'Wallet summary');
  }),

  transactions: asyncHandler(async (req: Request, res: Response) => {
    const query = parseListQuery(req);
    const { items, meta } = await walletService.transactions(req.user!.id, query);
    return ok(res, items, 'Wallet transactions', meta);
  }),

  topUp: asyncHandler(async (req: Request, res: Response) => {
    const amount = Number(req.body?.amount);
    const description = typeof req.body?.description === 'string' ? req.body.description : undefined;
    const data = await walletService.topUp(req.user!.id, amount, description);
    return ok(res, data, 'Top-up successful');
  }),
};
